addon/model.js

```diff
@@ -98,33 +112,76 @@ class YesManAttributesSingletonClass {
 
 const YesManAttributes = new YesManAttributesSingletonClass();
 
-const retrieveFromCurrentState = computed('_topModel.currentState', function(key) {
-  return this._topModel._internalModel.currentState[key];
-}).readOnly();
-
 // global buffer for initial properties to work around
 //  a)  can't write to `this` before `super`
 //  b)  core_object writes properties before calling `init`; this means that no
 //      CP or setknownProperty can rely on any initialization
 let initProperites = Object.create(null);
 
+let ignoreInvalidRequestsMap = new WeakMap();
+let ignoreErrorRequestsMap = new WeakMap();
 export default class MegamorphicModel extends EmberObject {
   init(properties) {
     // Drop Ember.Object subclassing instead
     super.init(...arguments);
+    if (false) {
+      recordDataToRecordMap.set(properties._recordData, this);
+      this._recordData = properties._recordData;
+      this._invalidRequests = [];
+      this._errorRequests = [];
+      this._lastError = null;
+    }
     this._store = properties.store;
-    this._internalModel = properties._internalModel;
     this._cache = Object.create(null);
     this._schema = get(properties.store, '_schemaManager');
 
     this._topModel = this._topModel || this;
     this._parentModel = this._parentModel || null;
     this._errors = DS.Errors.create();
     this._init = true;
+    if (!false) {
+      this._internalModel = properties._internalModel;
+    }
 
     this._flushInitProperties();
   }
 
+  _setIdentifier(identifier) {
+    if (false) {
+    }
+    this._identifier = identifier;
+    this.store.getRequestStateService().subscribeForRecord(this._identifier, request => {
+      if (request.state === 'rejected') {
+        // TODO filter out queries
+        this._lastError = request;
+        if (!(request.result && request.result.error instanceof DS.InvalidError)) {
+          this._errorRequests.push(request);
+        } else {
+          this._invalidRequests.push(request);
+        }
+      } else if (request.state === 'fulfilled') {
+        this._invalidRequests = [];
+        this._errorRequests = [];
+        this._lastError = null;
+      }
+      this._notifyNetworkChanges();
+    });
+  }
+
+  _notifyNetworkChanges() {
```

Comment:
let's change this to not fire more change notifications than we currently do (eg not firing `isSaving` change notification when rolling back errors)

Reply:
decided not to do this, because the tradeoff of making the code more complex, compared to couple more notifications doesn't seem worth it. Happy to change that if it causes problems, but the only way it would would be with observers